tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1280 @@
+/// let client1 = NamedPipeClientOptions::new().create(PIPE_NAME)?;
+///
+/// // this times out because the server is busy.
+/// let e = wait_named_pipe(PIPE_NAME, Some(Duration::from_millis(10))).await.unwrap_err();
+/// assert_eq!(e.kind(), io::ErrorKind::TimedOut);
+/// # Ok(()) }
+/// ```
+///
+/// # Examples
+///
+/// ```
+/// use std::io;
+/// use std::time::Duration;
+/// use tokio::io::{AsyncReadExt as _, AsyncWriteExt as _};
+/// use tokio::net::windows::{NamedPipeOptions, NamedPipeClientOptions, wait_named_pipe};
+/// use winapi::shared::winerror;
+///
+/// const PIPE_NAME: &str = r"\\.\pipe\tokio-named-pipe-client-wait";
+///
+/// # #[tokio::main] async fn main() -> std::io::Result<()> {
+/// // The first server needs to be constructed early so that clients can
+/// // be correctly connected. Otherwise calling .wait will cause the client
+/// // to error because the file descriptor doesn't exist.
+/// //
+/// // Here we also make use of `first_pipe_instance`, which will ensure
+/// // that there are no other servers up and running already.
+/// let mut server = NamedPipeOptions::new()
+///     .first_pipe_instance(true)
+///     .create(PIPE_NAME)?;
+///
+/// let client = tokio::spawn(async move {
+///     // Wait forever until a socket is available to be connected to.
+///     let mut client = loop {
+///         match NamedPipeClientOptions::new().create(PIPE_NAME) {
+///             Ok(client) => break client,
+///             Err(e) if e.raw_os_error() == Some(winerror::ERROR_PIPE_BUSY as i32) => (),
+///             Err(e) => return Err(e),
+///         }
+///
+///         wait_named_pipe(PIPE_NAME, Some(Duration::from_secs(5))).await?;
+///     };
+///
+///     let mut buf = [0u8; 4];
+///     client.read_exact(&mut buf[..]).await?;
+///     Ok::<_, io::Error>(buf)
+/// });
+///
+/// let server = tokio::spawn(async move {
+///     tokio::time::sleep(Duration::from_millis(200)).await;
+///
+///     // Calling `connect` is necessary for the waiting client to wake up,
+///     // even if the server is created after the client.
+///     server.connect().await?;
+///
+///     server.write_all(b"ping").await?;
+///     Ok::<_, io::Error>(())
+/// });
+///
+/// let (client, server) = tokio::try_join!(client, server)?;
+/// let payload = client?;
+/// assert_eq!(&payload[..], b"ping");
+/// let _ = server?;
+/// # Ok(()) }
+/// ```
+///
+/// # Panics
+///
+/// Panics if the specified duration is larger than `0xffffffff`
+/// milliseconds, which is roughly equal to 1193 hours.
+///
+/// ```should_panic
+/// use std::time::Duration;
+/// use tokio::net::windows::wait_named_pipe;
+///
+/// const PIPE_NAME: &str = r"\\.\pipe\mynamedpipe";
+///
+/// # #[tokio::main] async fn main() -> std::io::Result<()> {
+/// wait_named_pipe(PIPE_NAME, Some(Duration::from_millis(0xffffffff))).await?;
+/// # Ok(()) }
+/// ```
+pub async fn wait_named_pipe(addr: impl AsRef<OsStr>, timeout: Option<Duration>) -> io::Result<()> {
```

Comment:
I agree that we punt on this.

Reply:
I would probably want to include some version of it so that [people porting code to Tokio](https://github.com/search?q=WaitNamedPipe) has something to use without pulling in winapi, but not put it on the thread pool as suggested earlier.

It is the only function available for waking up a client once a socket becomes available ([despite being flawed before Windows 10](https://github.com/tokio-rs/tokio/pull/3760#issuecomment-839354959)) and not supporting overlapped I/O. The alternative is you sleep in a loop which also has its downsides.

Reply:
So it *is* still included, but is no longer `async` and all other examples have been updated to loop using a short `tokio::time::sleep` instead.

Reply:
So should this function be removed (with the above in mind)? Yay or nay?

It's almost like it's too much of a footgun to keep around.

Reply:
I've definitely used `WaitNamedPipe()` for sync apps, but it seems too likely someone is going to do `spawn_blocking(|| { wait_named_pipe(...); }).await` if it's there in tokio land, while the most likely if you don't is going to be the "correct" thing of looping.
I *might* suggest a helper that does this for you, given how likely the need is, but it's a little outside the general style of tokio AFAICT to provide "actual functionality" rather than just wrapping.

Reply:
Having a blocking version of it is also not great.